test/test_emulation/__init__.py

```diff
@@ -0,0 +1,168 @@
+import fcntl
+import libevdev
+import requests
+import shutil
+import subprocess
+import os
+import pytest
+import threading
+
+from pathlib import Path
+from time import sleep, time
+
+
+class RatbagemuClient(object):
+    def __init__(self, url='http://localhost:8080'):
+        self.url = url
+
+    '''
+    Internal wrappers around the http requests
+    '''
+    def _get(self, path):
+        return requests.get(f'{self.url}{path}')
+
+    def _post(self, path, data=None, json=None):
+        return requests.post(f'{self.url}{path}',
+                             data=data,
+                             json=json)
+
+    def _delete(self, path):
+        return requests.delete(f'{self.url}{path}')
+
+    def _put(self, path, data=None, json=None):
+        return requests.put(f'{self.url}{path}',
+                            data=data,
+                            json=json)
+
+    '''
+    ratbag-emu functions
+    '''
+    def create(self, shortname):
+        data = {
+            'shortname': shortname
+        }
+        response = self._post('/devices/add', json=data)
+        assert response.status_code == 201
+        return response.json()['id']
+
+    def delete(self, id):
+        response = self._delete(f'/devices/{id}')
+        assert response.status_code == 204
+
+    def get_dpi(self, id, dpi_id):
+        response = self._get(f'/devices/{id}/phys_props/dpi/{dpi_id}')
+        assert response.status_code == 200
+        return response.json()
+
+    def set_dpi(self, id, dpi_id, new_dpi):
+        response = self._put(f'/devices/{id}/phys_props/dpi/{dpi_id}', json=new_dpi)
+        assert response.status_code == 200
+
+    def get_input_nodes(self, id):
+        response = self._get(f'/devices/{id}')
+        if response.status_code == 200 and 'input_nodes' in response.json():
+            return response.json()['input_nodes']
+
+    def send_event(self, id, event):
+        response = self._post(f'/devices/{id}/event', json=event)
+        assert response.status_code == 200
+
+
+class TestDevice(object):
+    def reload_udev_rules(self):
+        subprocess.run('udevadm control --reload-rules'.split())
+
+    @pytest.fixture(scope='session', autouse=True)
+    def udev_rules(self):
+        rules_dir = Path('/run/udev/rules.d')
+        rules_dir.mkdir(exist_ok=True)
+        shutil.copyfile('test/test_emulation/rules.d/61-ratbag-emu-ignore-test-devices.rules', '/run/udev/rules.d/61-ratbag-emu-ignore-test-devices.rules')
+        self.reload_udev_rules()
+
+        yield
+
+        rules = rules_dir.joinpath('61-ratbag-emu-ignore-test-devices.rules')
+        if rules.is_file():
+            rules.unlink()
+            self.reload_udev_rules()
+
+    @pytest.fixture(autouse=True, scope='session')
+    def server(self):
+        p = None
+        try:
+            p = subprocess.Popen('python -m ratbag_emu 2&>1 >/dev/null', shell=True)
```

Comment:
see [popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen), you can route `stdin` etc to `subprocess.DEVNULL` rather than relying on a shell feature.